### PR TITLE
autoid_service: fix non-continuous autoid for allocating N IDs

### LINKED ISSUE #52465: after  create  table add auto_id_cache ,  the  AUTO_INCREMENT col have the huge gap
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
sysbench  1.20 
tidb  7.5.1
1)vim oltp_common.lua   modify create_table function add  AUTO_ID_CACHE=1
2) the create table  sql  like this:
CREATE TABLE sbtest2.sbtest4 (

id int(11) NOT NULL AUTO_INCREMENT,

k int(11) NOT NULL DEFAULT ‘0’,

c char(120) NOT NULL DEFAULT ‘’,

pad char(60) NOT NULL DEFAULT ‘’,

PRIMARY KEY (id) /*T![clustered_index] CLUSTERED */,

KEY k_4 (k)

) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin /*T![auto_id_cache] AUTO_ID_CACHE=1 */

time sysbench /usr/share/sysbench/oltp_read_write.lua --db-driver=mysql --mysql-host=xxx.xxx.16.65 --mysql-port=6000 --mysql-user=root --mysql-password=‘xxx’ --mysql-db=sbtest --table_size=1000000 --tables=10 --threads=10 --time=300 prepare
### 2. What did you expect to see? (Required)
the max(id)  from  sbtest_test = the count(*)  from sbtest_table

### 3. What did you see instead (Required)
select max(id),min(id) ,count(1) from sbtest2.sbtest4;

±--------±--------±---------+

| max(id) | min(id) | count(1) |

±--------±--------±---------+

| 1486752 | 1 | 1000000 |

±--------±--------±---------+
insert 2688 rows ,the max(id)=count(*),  
but  insert 2689 rows ,the max(id)=4001  ,count(*)=2689

### 4. What is your TiDB version? (Required)
7.5.1

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #52465

Problem Summary:

### What changed and how does it work?

When we require N autoids, and the current cached ids is not enough, we discard the remain ids and the allocation start with a new batch. This cause the "hole" in the result auto ids.

### Check List

Tests 

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/autoid_service/autoid.go [hunk 1]
```diff
@@ -149,7 +158,11 @@ func (alloc *autoIDValue) alloc4Signed(ctx context.Context,
 				return err1
 			}
 			// calcNeededBatchSize calculates the total batch size needed on global base.
-			n1 = calcNeededBatchSize(newBase, int64(n), increment, offset, isUnsigned)
+			if alloc.base == 0 {
```
Comment (Defined2014) [inline pkg/autoid_service/autoid.go:161]: When it will equals to `0`?

Comment (tiancaiamao) [reply pkg/autoid_service/autoid.go:161]: When the service is not initialized ... it need to get the value from remote.

**Review by Defined2014 — APPROVED** [review]

File: pkg/autoid_service/autoid.go [hunk 2]
```diff
@@ -168,7 +181,7 @@ func (alloc *autoIDValue) alloc4Signed(ctx context.Context,
 		if newBase == math.MaxInt64 {
 			return 0, 0, errAutoincReadFailed
 		}
-		alloc.base, alloc.end = newBase, newEnd
+		alloc.end = newEnd
 	}
 	min = alloc.base
 	alloc.base += n1
```
Comment (Defined2014) [inline pkg/autoid_service/autoid.go:187]: The `alloc.base` is incorrect. Maybe we needn't L147

**Review by Defined2014 — CHANGES_REQUESTED** [review]

**Review by Defined2014 — APPROVED** [review]

**Comment by tiancaiamao** [comment]
PTAL @bb7133

**Comment by tiancaiamao** [comment]
ping @bb7133

**Comment by tiancaiamao** [comment]
ping @bb7133

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**